### ISSUE TITLE
Quick Start Example Issue

### ISSUE THREAD

**Issue by angelosarto** [seawatts/ember-simple-auth-auth0#67]
I tried to create the simplest example possible and I ran across a few issues:

1. I started with a blank ember new app
1. I did ember install ember-simple-auth-auth0
1. I added the route, controller and template code
1.  The lock component loads, and I can login, after a successful login the url contains the JWT
localhost:4200/#access_token=ey...
however I the session doesn't seem to be present in the template and there is an error `Error while processing route: index` in the console

1.  However, if I refresh the page the session data is displayed. (email address).   
1.  If I delete the token from the URL it still works (I imagine because it is now in the store)

(I have now tried this with the latest ember-simple-auth-auth0)

AM I missing something?

**Comment by philipsoutham** [comment]
@angelosarto I was experiencing the same thing. In `app/controllers/application.js` I added the `responseType` attribute to my `lockOptions` so it looks like:

```javascript
const lockOptions = {
  auth: {
    params: {
      scope: 'openid user_metadata',
      responseType: 'id_token token',
    }
  }
};
```
Now it works without the refresh.

**Comment by jmar910** [comment]
None of this is working for me. What im noticing is that auth0 is bubbling up an error saying that the `Nonce does not match` which is causing the `Error while processing route` but when I hard refresh the callback route like @angelosarto, it works fine. I've been hunting down the issue all last night, but still have not turned up why this is happening.

seems related to https://github.com/auth0/auth0.js/issues/365 and I saw that @seawatts makes an appearance in the comments, but im also noticing other weird stuff, like the `beforeModel` hook in the mixin being called twice. In certain cases, the first time its called, it passes verification with the nonce, but then it gets called a second time and the nonce doesn't match, causing an error on the route.

I'm going to continue digging but I'm starting to pull my hair out over the issue. 😢

**Comment by jmar910** [comment]
Okay so I may have been shooting myself in the foot.... although im still a little confused. I moved the mixin to only be on the callback route and that worked. I suppose beforeModel was getting called twice, once when I entered my callback route, and then again when I tried to redirect to my `protected` route.

This isn't a good solution obviously because it breaks logout. I'm wondering if maybe the applicationroutemixin needs to be changed in this repo 🤔

**Comment by jmar910** [comment]
Alright so I copied the mixin into my project, and added this into the `beforeModel` right after the `promise` variable assignment to make it all work. I'm not thrilled about the solution, but it does prevent the `beforeModel` hook in the mixin from being called twice, and auth is working so :man_shrugging: 

```
if (transition.targetName !== 'auth-callback') { //name of callback route
      return promise;
}
```

From what I could tell, ESA's `sessionAuthenticated()` hook is firing, causing a transition into whatever you configure as your `routeAfterAuthentication` which is causing the `beforeModel` in the mixin to be fired again.

**Comment by sukima** [comment]
Is there any official answer on this problem. I find it hard to believe you and I are the only ones with this problem and the fix seems to be a bit over the top for an addon. It doesn't make sense in that way.

Id really like to get more information on why others are not experiencing this issue. This has been a blocker for my home projects for about 3 weeks now. I've essentially given up on Auth0 because of it. @seawatts / @jmar910 can we make any movement on this at all? Hints on possible PRs or better workarounds? Auth0 can't be dead in the water for Ember folks!

**Comment by sukima** [comment]
Also I think the title should be changed because this isn't a documentation problem it's an `ApplicationMixin` **bug**.

**Comment by jmar910** [comment]
@sukima i'm curious, does my workaround work for you? I'd say use that for now, and I'll try to find time to see if i can whip up a real fix. Ive been using it and haven't given it a second thought. it might not be pretty but it holds up. 

I am also curious how others are not running into this, or maybe they are and it's not being reported?

**Comment by seawatts** [comment]
Yeah I think the issue is that it is being called twice. I need to fix this.

**Comment by jmar910** [comment]
@seawatts sweet, let me know if i can help

**Comment by sukima** [comment]
Does the Auth0 JS library offer a promise instead of hijacking the URL?

**Comment by deeksy** [comment]
OK, just been working through this one myself.  The issue seems to occur if a nonce is created by the auth0 lock.js container, which only occurs for me when using a social provider.  If I don't use a social provider, then no nonce is created by auth0, and no nonce checking happens.

The way I worked around this was by modifying the _getNewUrlHashData() method so that it explicitly passes in the window.location.hash object to parseHash, like follows:
```js
    return new Ember.RSVP.Promise((resolve, reject) => {
        auth0.parseHash(window.location.hash, (err, parsedPayload) => {
          if (err) {
            console.log('got an error parsing the harsh', err);
            if (err.errorDescription) {
              err.errorDescription = decodeURI(err.errorDescription);
            }

            return reject(err);
          }

          resolve(parsedPayload);
        });
      });
```

Once I made that change, auth0 correctly parsed the hash from the url, and as the nonce matches the one created as part of the login process, worked correctly.

**Comment by sukima** [comment]
I gave up trying and simply used [torii](http://vestorly.github.io/torii/) to open up my Auth0's OAuth URL. Old school but means I can remove the auth0 lock.js which is not fastboot friendly.

**Comment by medokin** [comment]
So this Issue makes this addon unusable?

**Comment by ioanszabo** [comment]
Hi, What I did in order to inspect not to fix beforeModel, I added beforeModel in application route like this:
```
beforeModel() {
	this._super(...arguments)
},
```
Strange, but it worked as expected without doing nothing. access_token and other parameters disappeared from url and I was redirected to my good dashboard page.

**Comment by rzwink** [comment]
I took ioanszabo's advice, added 'this' to the parameters, and my nonce mismatch issue went away.
routes/application.js

```
  beforeModel() {
    this._super.apply(this, ...arguments);
  },

```

**Comment by XaserAcheron** [comment]
Just as a random update, I was setting up a sample project earlier and ran into this exact issue (as described in the first post). I'll take a look at this as soon as I can, but the immediate priority is cutting a release.

**Comment by peavers** [comment]
I was having the exact issues everyone else here was having, however after manually clearing my browser local storage, the example app started working as expected.

**Comment by nickschot** [comment]
I am also encountering this. Refreshing when the token stuff is in the URL always makes it break for me. Removing it from the URL works fine. I've tried the proposed fixes, but they won't work for me unfortunately.

**Comment by XaserAcheron** [comment]
If it helps, I get this behavior when trying to use "redirect mode;" using popup mode (archaic as it is) works fine, for folks who find it a feasible workaround (YMMV).

**Comment by XaserAcheron** [comment]
I've finally did some proper digging on this. Turns out that refreshing the page with the token in the URL causes Auth0 to throw an error with the following payload:
```
{
  error: "invalid token",
  errorDescription: "`state` does not match"
}
```
... which is actually totally legit, sorta. That's Auth0's automatic [CSRF mitigation](https://auth0.com/docs/protocols/oauth2/oauth-state) in action.

There's basically two issues at play here:
- The error details are being swallowed during promise rejection (eep!), leading to an unhelpful "Error while processing route: index" whitescreen.
- The URL hash isn't properly cleaned up after it's used to authenticate, which is the real root cause of all this mess.

I've fixed the error details for sanity's sake, and I'm currently working on the hash cleanup stuff. Sorry it took me so long to get 'round to it.
